<a href="https://colab.research.google.com/github/amonguscline/ML-Data-analysis/blob/main/notes/Random_Forest.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression

In [ ]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/abalone/abalone.data"
df=pd.read_csv(url)
df.columns=['Sex','Length','Diameter','Height','Whole weight', 'Shucked weight', 'Viscera weight', 'Shell weight', 'Rings']
y=df["Rings"]
X=df.iloc[:,:-1]
X=pd.get_dummies(X, dtype=float)
X_train,X_test,y_train,y_test=train_test_split(X,y)

In [ ]:
model=RandomForestRegressor(max_depth=9, n_estimators=400,
                            max_features=.60, min_samples_leaf=4,
                            max_samples=.8) #n_estimators is the batch of decision trees in the forest, max features is the number of rows accounted for in split
model.fit(X_train,y_train)
model.score(X_test,y_test)

0.5610635929014787

In [ ]:
model=LinearRegression()
model.fit(X_train,y_train)
model.score(X_test,y_test)

0.5308666504372116

#Cross validation


testing the machine learning model on training data is bad because of overfitting, and splitting the data into training and testing data can be good but it can lead to different test sets. To solve this, we can use cross validation, which splits the data into multiple sets and then we can get multiple scores based on those splits. cross validation also makes sure that each data point is used in different splits in different times. we use the average score of all splits, which will give a better estimate of how the model performs in the real world. we usually use 3,5,10 or 20 depending on the size of the data.

In [ ]:
from sklearn.model_selection import cross_val_score

In [ ]:
model = RandomForestRegressor(max_depth=8, n_estimators=80,
                            max_features=.7, min_samples_leaf=5,
                            max_samples=.8) #max_features forces more variation, same with max_samples
(-cross_val_score(model,X,y,cv=5,scoring='neg_mean_squared_error').mean())**(1/2)

2.2021931678535047

The RMSE is the error of what you're trying to predict on average. Here, the number of rings should be off by about 2.20 on average. RMSE is used a lot more in regression

In [ ]:
y.describe() #we are beating the standard deviation

,Rings
count,4176.000000
mean,9.932471
std,3.223601
min,1.000000
25%,8.000000
50%,9.000000
75%,11.000000
max,29.000000


In [ ]:
model.fit(X_train,y_train)
model.predict(X_test.tail())

array([10.82463737, 11.2886039 , 10.50478078,  8.11531771, 10.19911513])

In [ ]:
y.value_counts() #maybe cut outliers 3<y<23
df=df[(df["Rings"]>2) & (df["Rings"]<24)]
df["Rings"].value_counts()
y=df["Rings"]
X=df.iloc[:,:-1]
X=pd.get_dummies(X, dtype=float)
(-cross_val_score(model,X,y,scoring='neg_mean_squared_error').mean())**(1/2)

2.159365528239482